### PR TITLE
various loading time optimization

### PR DESCRIPTION
**Pull request by Glease**
no more stinky registerUsagesForMaterials yay! and I did it without parallelism!

### CODE REVIEW

File: docs/structure_code.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# GT5U Structure Code Guide
```
Comment (eigenraven) [inline docs/structure_code.md:1]: Did you include this file by mistake?

File: src/main/java/gregtech/api/util/GT_RecipeRegistrator.java [hunk 2]
```diff
@@ -4,17 +4,25 @@
...
 import gregtech.GT_Mod;
 import gregtech.api.GregTech_API;
 import gregtech.api.enums.*;
 import gregtech.api.objects.ItemData;
 import gregtech.api.objects.MaterialStack;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.List;
+import ic2.api.reactor.IReactorComponent;
+import java.util.*;
```
Comment (leagris) [inline src/main/java/gregtech/api/util/GT_RecipeRegistrator.java:14]: There need to be some clean-up with imports.

File: docs/structure_code.md (hunk 1 above)
Comment (Dream-Master) [reply docs/structure_code.md]: i removed it

Comment (Glease) [reply docs/structure_code.md]: bruh yeah it shouldn't be there. Thanks for the help.

File: src/main/java/gregtech/api/util/GT_RecipeRegistrator.java (hunk 2 above)
Comment (Glease) [reply src/main/java/gregtech/api/util/GT_RecipeRegistrator.java:14]: spotless complains if you place java in a separate block

File: src/main/java/gregtech/api/util/GT_RecipeRegistrator.java [hunk 3]
```diff
@@ -445,6 +459,102 @@ public static synchronized void registerUsagesForMaterials(
...
+            x.clear();
+            ItemStack tStack = tRecipe.getRecipeOutput();
+            if (GT_Utility.isStackValid(tStack)
+                    && tStack.getMaxStackSize() == 1
+                    && tStack.getMaxDamage() > 0
+                    && !(tStack.getItem() instanceof ItemBlock)
+                    && !(tStack.getItem() instanceof IReactorComponent)
+                    && !GT_ModHandler.isElectricItem(tStack)
+                    && !GT_Utility.isStackInList(tStack, GT_ModHandler.sNonReplaceableItems)) {
+                if (tRecipe instanceof ShapelessRecipes || tRecipe instanceof ShapelessOreRecipe) {
```
Comment (eigenraven) [inline src/main/java/gregtech/api/util/GT_RecipeRegistrator.java:494]: Move this if before the tStack checks, instanceof will be quicker than scanning lists

**Review by eigenraven — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/api/util/GT_RecipeRegistrator.java (hunk 3 above)
Comment (Glease) [reply src/main/java/gregtech/api/util/GT_RecipeRegistrator.java:494]: moved.

**Review by eigenraven — APPROVED** [review]

**Comment by Glease** [comment]
Just come to think about this, wouldn't this also be a good way to speed up vanilla crafting bench recipe matching and not just GT loading?

**Comment by eigenraven** [comment]
Probably, you could mixin it via hodgepodge into the shapedrecipe implementation I guess.

**Pull request merged**